certbot/lock.py

```diff
@@ -108,17 +164,101 @@ def _lock_success(self, fd):
         try:
-            compat.release_locked_file(self._fd, self._path)
+            os.remove(self._path)
         finally:
-            self._fd = None
+            try:
+                if self._fd:
+                    os.close(self._fd)
+            finally:
+                self._fd = None
+
+    def is_locked(self):
+        # type: () -> bool
+        """Check if lock file is currently locked.
+        :return: True if the lock file is locked
+        :rtype: bool
+        """
+        return self._fd is not None
+
+
+# Windows locking mechanism
+# ~~~~~~~~~~~~~~~~~~~~~~~~~
+class _WindowsLockMechanism(object):
+    """
+    A Windows lock file mechanism.
+    On Windows in general, access to a file is exclusive, so opening a file
+    is an effective lock. This default behavior may be modified by an administrator,
+    so we ensure correct behavior with msvcrt. Moreover on Windows a file can be removed
+    only after it has been released: so the concurrency access that may occur on POSIX
+    system is irrelevant here, leading to a quite simple code.
+    """
+
+    def __init__(self, path):
+        # type: (str) -> None
+        self._path = path
+        self._fd = None
+
+    def acquire(self):
+        """Acquire the lock"""
+        open_mode = os.O_RDWR | os.O_CREAT | os.O_TRUNC
+
+        fd = os.open(self._path, open_mode, 0o600)
+        try:
+            msvcrt.locking(fd, msvcrt.LK_NBLCK, 1)
+        except (IOError, OSError) as err:
+            # Anything except EACCES is unexpected. Raise directly the error in that case.
+            if err.errno != errno.EACCES:
+                raise
+            os.close(fd)
+            _raise_for_certbot_lock(self._path)
+
+        self._fd = fd
+
+    def release(self):
+        """Release the lock."""
+        if self._fd:
+            try:
+                msvcrt.locking(self._fd, msvcrt.LK_UNLCK, 1)
+                os.close(self._fd)
+
+                try:
+                    os.remove(self._path)
+                except OSError as e:
+                    # If the lock file cannot be removed, it is not a big deal.
+                    # Likely another instance is acquiring the lock we just released.
+                    logger.debug(str(e))
+            finally:
+                self._fd = None
+
+    def is_locked(self):
+        # type: () -> bool
+        """Check if lock file is currently lock.
+        :return: True if the lock file is locked
+        :rtype: bool
+        """
+        return self._fd is not None
+
+
+def _raise_for_certbot_lock(lock_file_path):
+    # type: (str) -> None
+    """
+    Raise a certbot error when lock file cannot be acquired.
+    :param str lock_file_path: the path to the lock file
+    """
+    raise errors.LockError(
+        'Error, the lock file "{0}" could not be acquired. '
+        'It is likely that another Certbot instance is still running.'
+        .format(lock_file_path))
```

Comment:
I saw this comment before ^^

Agree, it is fare assumption that this exception is raised only with another Certbot currently running in parallel. Extremely unlikely that another process try to bother with locks we created.

Message updated. Also, for one line I think we can support the duplication of LOC instead of opening a brand new function. I move the exception in relevant places + debug logging.